EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs

```diff
@@ -0,0 +1,375 @@
+﻿using EngineLayer.FdrAnalysis;
+using Microsoft.ML;
+using Microsoft.ML.Data;
+using Proteomics.ProteolyticDigestion;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using static Microsoft.ML.DataOperationsCatalog;
+
+namespace EngineLayer
+{
+    public static class PEP_Analysis
+    {
+        public static string ComputePEPValuesForAllPSMsGeneric(List<PeptideSpectralMatch> psms)
+        {
+            Dictionary<string, int> accessionAppearances = GetAccessionCounts(psms);
+            Dictionary<string, int> sequenceToPsmCount = GetSequenceToPSMCount(psms);
+
+            MLContext mlContext = new MLContext();
+            IDataView dataView = mlContext.Data.LoadFromEnumerable(CreatePsmData(psms, accessionAppearances, sequenceToPsmCount));
+
+            //
+            // Summary:
+            //     Split the dataset into the train set and test set according to the given fraction.
+            //     Respects the samplingKeyColumnName if provided.
+            //
+            // Parameters:
+            //   data:
+            //     The dataset to split.
+            //
+            //   testFraction:
+            //     The fraction of data to go into the test set.
+            //
+            //   samplingKeyColumnName:
+            //     Name of a column to use for grouping rows. If two examples share the same value
+            //     of the samplingKeyColumnName, they are guaranteed to appear in the same subset
+            //     (train or test). This can be used to ensure no label leakage from the train to
+            //     the test set. If null no row grouping will be performed.
+            //
+            //   seed:
+            //     Seed for the random number generator used to select rows for the train-test split.
+            TrainTestData trainTestSplit = mlContext.Data.TrainTestSplit(dataView, testFraction: 0.1);
+            IDataView trainingData = trainTestSplit.TrainSet;
+            IDataView testData = trainTestSplit.TestSet;
+
+            var trainer = mlContext.BinaryClassification.Trainers.FastTree(labelColumnName: "Label", featureColumnName: "Features");
+
+            var pipeline = mlContext.Transforms.Concatenate("Features", "Intensity", "ScanPrecursorCharge", "DeltaScore", "Notch", "PsmCount", "ModsCount", "MissedCleavagesCount", "Ambiguity", "AccessionAppearances", "LongestFragmentIonSeries")
```

Comment:
Including PSM count as a ML parameter is dangerous and inherently biased for the identification of targets over decoys, such that a false-positive target is more likely to be identified than a decoy.

From [this paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4751987/):
"One of the better documented problems with TD estimation of FDR is its incompatibility with any scoring system that uses protein inference, such as weighting based on the presence of other peptides from the same protein. As more peptides are being matched to the target database, it is much more likely that a random PSM to the target database is to a protein already identified than a decoy PSM being to the same protein as another decoy database match. It was largely for this reason that the Percolator software removed protein inference parameters from its re-scoring."

Reply:
I would not feel comfortable using PEPs until that parameter is removed.

Reply:
there is some potential for bias. however. the training set for decoy is mixed with low scoring trues. i dont have the actual number in from but id say the 40% of the false category is of such time (60% actual decoys). The damage is mitigated somewhat. Also, i did a search of mouse data and added arabadopsis database. After, I compared arabaopsis true to arabadopsis decoy in the results. A ranked set, showed that the two occured with basically the same frequency across the score range. This particular decoy/decoy search also shows that ther is little if any bias. though i'm sure there must be some

Reply:
There is not a good way to measure this bias. Low scoring targets used for training are identified purely from MetaMorpheus score and largely have PSM counts of 1. The consideration of PSM scores allow for those with randomly higher PSM counts (e.g. 2) to be unfairly increased. Entrapment with Arabidopsis won't work to measure this bias, because Arabidopsis proteins aren't present in the mouse sample to unfairly bias the Arabidopsis target sequences, only the mouse target sequences.
Independent of the previously mentioned bias, this parameter is somewhat useless for protein identification and quantification. Although it may results in more PSMs (arguably because the FDR is underestimated), the number of proteins and the number of quantified peptides should not change, because this parameter only favors the identification of PSMs that were already identified. As such, it's more for posturing (we got more PSMs) than for actual benefit.

Reply:
This parallels the issue with the classic method for protein FDR (not top-picked). In the classic, we identify a large number of target proteins before we start hitting false-positives. When we hit a false positive, there's no longer a 50/50 chance of hitting a target/decoy, because targets have been removed from the space. This allows decoys a higher probability of identification and results in an overestimated FDR. The opposite problem occurs here when using the PSM count parameter. We identified a high number of target PSMs that were actually present in the sample. When we get to a false-positive, there is no longer a 50/50 shot of a target/decoy being matched because identifications are biased toward the targets that were previously identified, resulting in an underestimated FDR.